edit/source-editor.js

```diff
@@ -0,0 +1,265 @@
+@name New Style - ${Date.now()}
+@namespace github.com/openstyles/stylus
+@version 0.1.0
+@description A new userstyle
+@author Me
+==/UserStyle== */
+
+${section}
+`;
+    dirty.modify('source', '', sourceCode);
+    style.sourceCode = sourceCode;
+  }
+
+  function initHooks() {
+    // sidebar commands
+    $('#save-button').onclick = save;
+    $('#beautify').onclick = beautify;
+    $('#keyMap-help').onclick = showKeyMapHelp;
+    $('#toggle-style-help').onclick = showToggleStyleHelp;
+    $('#cancel-button').onclick = goBackToManage;
+
+    // enable
+    $('#enabled').onchange = e => {
+      const value = e.target.checked;
+      dirty.modify('enabled', style.enabled, value);
+      style.enabled = value;
+    };
+
+    // source
+    cm.on('change', () => {
+      const value = cm.getValue();
+      dirty.modify('source', style.sourceCode, value);
+      style.sourceCode = value;
+
+      updateLintReportIfEnabled(cm);
+    });
+
+    // hotkeyRerouter
+    cm.on('focus', () => {
+      hotkeyRerouter.setState(false);
+    });
+    cm.on('blur', () => {
+      hotkeyRerouter.setState(true);
+    });
+
+    // autocomplete
+    if (prefs.get('editor.autocompleteOnTyping')) {
+      setupAutocomplete(cm);
+    }
+  }
+
+  function updateMetas() {
+    $('#name').value = style.name;
+    $('#enabled').checked = style.enabled;
+    $('#url').href = style.url;
+    const {usercssData: {preprocessor} = {}} = style;
+    cm.setPreprocessor(preprocessor);
+    // beautify only works with regular CSS
+    $('#beautify').disabled = cm.getOption('mode') !== 'css';
+    updateTitle();
+  }
+
+  function updateTitle() {
+    // title depends on dirty and style meta
+    if (!style.id) {
+      document.title = t('addStyleTitle');
+    } else {
+      document.title = (dirty.isDirty() ? '* ' : '') + t('editStyleTitle', [style.name]);
+    }
+  }
+
+  function replaceStyle(newStyle) {
+    if (!style.id && newStyle.id) {
+      history.replaceState({}, '', `?id=${newStyle.id}`);
+    }
+    style = deepCopy(newStyle);
+    updateMetas();
+    if (style.sourceCode !== cm.getValue()) {
+      const cursor = cm.getCursor();
+      cm.setValue(style.sourceCode);
+      cm.setCursor(cursor);
+    }
+    dirty.clear();
+    hadBeenSaved = false;
+  }
+
+  function toggleStyle() {
+    const value = !style.enabled;
+    dirty.modify('enabled', style.enabled, value);
+    style.enabled = value;
+    updateMetas();
+    // save when toggle enable state?
+    save();
+  }
+
+  function save() {
+    if (!dirty.isDirty()) {
+      return;
+    }
+    return onBackgroundReady()
+      .then(() => BG.usercssHelper.save({
+        reason: 'editSave',
+        id: style.id,
+        enabled: style.enabled,
+        sourceCode: style.sourceCode
+      }))
+      .then(replaceStyle)
+      .then(() => {
+        hadBeenSaved = true;
+      })
+      .catch(err => {
+        console.error(err);
+        alert(err);
+      });
+  }
+
+  function isWarm() {
```

Comment:
I think isWarm is a deceptive name: I would expect it means the editor is loaded and ready to be used.
It's not clear why a separate function is needed in addition to isDirty.

Reply:
How about `isTouched`? The difference:

1. When a file is opened:
	* isDirty: false
	* isTouched: false
2. Edit the file:
	* isDirty: true
	* isTouched: true
3. Save the file:
	* isDirty: false
	* isTouched: true

Reply:
isTouched seems a better term (`touch` being a utility in *nix for a similar task). My question about why we need isTouched remains.

Reply:
It is used to determine whether the style could be reloaded without warning:

1. The user open a style in the editor.
2. Make some changes and save (`isDirty === false`, `isTouched  === true`).
3. The style is updated outside of the editor (e.g. from updater or live reload).
4. The editor should ask the user if they want to reload the style (drop the changes). If not, the editor should be set to dirty again.

If `isTouched === false` (no change comparing to the original code), the style would be reloaded without prompting the user.

Reply:
make sense

Reply:
makes sense